packages/app/src-tauri/src/commands/s3.rs

```diff
@@ -1,100 +1,249 @@
+    let file_payload = FilePayload {
+        local_path: temp_file.path().to_owned(),
+        connection: payload.connection,
+        endpoint_config: payload.endpoint_config,
+    };
+
+    Ok(s3_upload_progress_internal(
+        channel,
+        file_payload,
+        Some(temp_file),
+        use_path_style,
+    ))
+}
+
+#[named]
+fn s3_upload_progress_internal(
+    channel: Channel,
+    payload: FilePayload,
+    temp_file: Option<TempFile>,
+    use_path_style: bool,
+) -> Uuid {
+    let mut event_producer = EventProducer::new(channel);
+    let id = event_producer.id();
+
+    spawn(async move {
+        let result =
+            s3_upload_progress_task(&mut event_producer, payload, temp_file, use_path_style)
+                .await
+                .map(Event::FileResult);
+        log_if_error_and_return!(&result);
+        event_producer.send(result).await;
+    });
+    id
+}
+
+async fn s3_upload_progress_task(
+    event_producer: &mut EventProducer,
+    payload: FilePayload,
+    _temp_file: Option<TempFile>,
+    use_path_style: bool,
+) -> Result<FileInfo> {
+    // Init Phase
+    event_producer.send(Ok(Event::Progress(0))).await;
+
+    let mut bucket = payload.connection.bucket()?;
+
+    // this header is required to make the uploaded file public readable.
+    bucket.add_header("x-amz-acl", "public-read");
+
+    //Useful for testing/old S3 implementation
+    if use_path_style {
+        bucket.set_path_style();
+    }
+
+    event_producer.send(Ok(Event::DeltaProgress(2))).await;
+
+    let ext = Path::new(&payload.local_path)
+        .extension()
+        .map_or(Cow::default(), |ext| ext.to_string_lossy());
+
+    let path = payload
+        .endpoint_config
+        .path()
+        .as_ref()
+        .unwrap_or(&"".to_owned())
+        .to_owned();
+    let new_file_name: String = format!(
+        "{}/{}.{}",
+        &path,
+        &payload.endpoint_config.file_name(),
+        &ext
+    );
+
+    let file_info = get_file_info(&payload.local_path).await?;
+
+    event_producer.send(Ok(Event::DeltaProgress(5))).await;
+
+    // Transfer phase: 80-88%
+
+    // Step by 8%
+    let file_iter =
+        FileIter::new(payload.local_path.as_ref())?.set_mode(get_chunk::ChunkSize::Percent(10.));
+
+    let upload_response = bucket
+        .initiate_multipart_upload(&new_file_name, &file_info.mime_type)
+        .await?;
+
+    let mut set: JoinSet<Result<()>> = JoinSet::new();
+
+    for (index, chunk) in file_iter.enumerate() {
+        // Prepare data for the part upload task
+        let chunk = chunk?;
+        let mut event_producer = event_producer.clone();
+        let bucket = bucket.clone();
+        let new_file_name = new_file_name.clone();
+        let upload_id = upload_response.upload_id.clone();
+        let mime_type = file_info.mime_type.clone();
+
+        set.spawn(async move {
+            bucket
+                .put_multipart_chunk(chunk, &new_file_name, index as u32, &upload_id, &mime_type)
```

Comment:
Part number should start from 1, don't ask me why :D but seems that this is a specification for making it working

```
  .put_multipart_chunk(
                    chunk,
                    &new_file_name,
                    index as u32 + 1, // S3 parts start at 1
                    &upload_id,
                    &mime_type,
```

Reply:
Fixed using map (index, chunck)  -->  (index + 1, chunck)